core/src/main/java/org/elasticsearch/common/io/stream/Streamable.java

```diff
@@ -22,11 +22,26 @@
 import java.io.IOException;
 
 /**
+ * Implementers can be written to a {@linkplain StreamOutput} and read from a {@linkplain StreamInput}. This allows them to be "thrown
+ * across the wire" using Elasticsearch's internal protocol. If the implementer also implements equals and hashCode then a copy made by
+ * serializing and deserializing must be equal and have the same hashCode. It isn't required that such a copy be entirely unchanged. For
```

Comment:
It would be great if we could enforce this contract somehow, e.g. add a test base class which all `Streamable` classes have to implement (but I think we'd be required to add lots of tests just for that). Wdyt?